gector/gec_predictor.py

```diff
@@ -143,15 +157,102 @@ def predict_batch_instance(self, instances: List[Instance]) -> List[JsonDict]:
+            # we will not correct it.
+            # It is directly written to output.
+            if len(instance.fields["tokens"].tokens) < 4:
+                final_outputs[id] = {
+                    "logits_labels": None,
+                    "logits_d_tags": None,
+                    "class_probabilities_labels": None,
+                    "class_probabilities_d_tags": None,
+                    "max_error_probability": None,
+                    "words": instance.fields["tokens"].tokens[1:],
+                    "labels": None,
+                    "d_tags": None,
+                    "corrected_words": instance.fields["tokens"].tokens[1:],
+                }
+            else:
+                pred_ids.append(id)
+
+        # Applying correction model multiple times
+        for _ in range(self._iterations):
+
+            # If no sentences need to be passed into model
+            if len(pred_ids) == 0:
+                break
+
+            # Create batch of instances to be passed into model
+            orig_batch = [final_batch[pred_id] for pred_id in pred_ids]
+
+            # Pass into model
+            outputs = self._model.forward_on_instances(orig_batch)
+
+            new_pred_ids = []
+
+            # Output_ID and Pred_ID in pred_ids
+            for op_ind, pred_id in enumerate(pred_ids):
+
+                # Update final outputs
+                final_outputs[pred_id] = outputs[op_ind]
+                orig = final_batch[pred_id]
+
+                # Create tokens from corrected words for next iter
+                tokens = [
+                    Token(word)
+                    for word in ["$START"] + outputs[op_ind]["corrected_words"]
+                ]
+
+                # Tokens to instance
+                pred = self._dataset_reader.text_to_instance(tokens)
+                prev_preds = prev_preds_dict[pred_id]
+
+                # If model output is different from previous iter outputs
+                # Update input batch, append to dict and add to `pred_ids`
+                if (
+                    orig.fields["tokens"].tokens != pred.fields["tokens"].tokens
+                    and pred.fields["tokens"].tokens not in prev_preds
+                ):
+                    # Set sentence instance to predicted sent in batch
+                    final_batch[pred_id] = pred
+                    # Append orig_id in new_pred_ids
+                    new_pred_ids.append(pred_id)
+                    # Append new prediction to prev_preds_dict
+                    prev_preds_dict[pred_id].append(
+                        pred.fields["tokens"].tokens
+                    )
+                # If model output is same as that in prev iter, update final batch
+                # but stop passing it into the model for future iters
+                # This means that no corrections have been made in this iteration
+                elif (
+                    orig.fields["tokens"].tokens != pred.fields["tokens"].tokens
```

Comment:
Oh I see so there was also an "early stopping" component as well where it stopped iterating if no corrections had been made. That's probably also the source of some of the discrepancies.

Reply:
Exactly! Passing it into the model again seemed to force it to make unnecessary corrections. That's where all the weird sentences I showed you in our meeting came from.